### ISSUE TITLE
Understanding Boomerang forward queries

### ISSUE THREAD

**Issue by copumpkin** [CROSSINGTUD/SPDS#91]
Hi! I put together a minimal example to illustrate my confusion:

```
import java.util.LinkedList;
import java.util.List;

class Foo {
	void bar() {
		System.out.println("zomg bar");
	}
}

public class Test {
	public static List<Foo> foos() {
		List<Foo> x = new LinkedList<>();
		Foo foo = new Foo();
		x.add(foo);
		foo.bar();
		return x;
	}

	public static void main(String[] args) {
		System.out.println("hi!");
		System.out.println(foos());
	}
}
```

Ignoring all the usual soot setup machinery (I'm using `DefaultBoomerangOptions` and the pretransformer), after compiling the code I end up with the following Jimple:

```java
...
        $stack2 = new java.util.LinkedList;
        specialinvoke $stack2.<java.util.LinkedList: void <init>()>();
        l0 = $stack2;
        $stack3 = new Foo;
        specialinvoke $stack3.<Foo: void <init>()>();
        l1 = $stack3;
        interfaceinvoke l0.<java.util.List: boolean add(java.lang.Object)>(l1);
        virtualinvoke l1.<Foo: void bar()>();
...
```

Now, if I point my forward query at `$stack2 = new java.util.LinkedList`, pass it to my solver instance, and call `getInvokedMethodOnInstance` on the result, I get the following:

```
{CS: l0.add(l1)=<java.util.List: boolean add(java.lang.Object)>, CS: $stack2.<init>()=<java.util.LinkedList: void <init>()>}
```

Which is what I'd expect, since I call the constructor on that list and also use `.add` on it.

On the other hand, if I do the exact same thing to the `$stack3 = new Foo` statement, I get an empty set of invoked methods, even though I call the constructor on it (obviously) and also call `.bar` on it.

I'm trying to understand what the fundamental difference is between those two cases and how I can make the latter query on my `Foo` instance also return the correct set of invoked methods.

**Comment by copumpkin** [comment]
Highlighting @johspaeth since I think you're most likely to know the answer and whether I'm doing anything wrong 😄

**Comment by johspaeth** [comment]
Hi @copumpkin,

thanks for tagging me and the question. 

The issue may be related to the way Soot is setup and the call graph is constructed. I assume that in your current setup `Foo` is not on your Soot classpath / explicitly loaded before the call graph is constructed (Probably similar to the test code setup). Then there will be no call graph edge from `<init>()` to the constructor and `virtualinvoke l1.<Foo: void bar()>();` to `bar()` present. 

How far does the data-flow of `$stack3` reach in your example? Does it propagate to `l1@virtualinvoke l1.<Foo: void bar()>()` or does the data-flow stop earlier? Because the call to `<init>` cannot be resolved in the call graph?

**Comment by copumpkin** [comment]
Thank you for your answer! In this case I believe `Foo` should be on the classpath because I basically just called `javac` on the file above and stuck the results into a jar and passed that in. So if `Test` is visible to Boomerang I'd expect `Foo` to be too. If this behavior is unexpected, I can give you a more complete repro down to the Soot initialization code, I just assumed I was probably doing something wrong.

> How far does the data-flow of $stack3 reach in your example? Does it propagate to l1@virtualinvoke l1.<Foo: void bar()>() or does the data-flow stop earlier? Because the call to <init> cannot be resolved in the call graph?

I'm not sure I'm following this question. Based on my "human analysis", the lifetime and dataflow of that instance is restricted to these three lines:

```
		Foo foo = new Foo();
		x.add(foo);
		foo.bar();
```

But in the code I don't really do anything with that `Foo` instance other than `println` it (i.e., `.toString()`) later on.

**Comment by johspaeth** [comment]
> Thank you for your answer! In this case I believe `Foo` should be on the classpath because I basically just called `javac` on the file above and stuck the results into a jar and passed that in. So if `Test` is visible to Boomerang I'd expect `Foo` to be too. If this behavior is unexpected, I can give you a more complete repro down to the Soot initialization code, I just assumed I was probably doing something wrong.

Soot has many different setup options, which need to be set properly and could go wrong in this scenario. The first thing to check is whether the call graph has been computed correctly. Are the edges to `bar` and to `Foo.<init>` present?


> > How far does the data-flow of $stack3 reach in your example? Does it propagate to l1@virtualinvoke l1.<Foo: void bar()>() or does the data-flow stop earlier? Because the call to  cannot be resolved in the call graph?
> 
> I'm not sure I'm following this question. Based on my "human analysis", the narrow lifetime and dataflow of that instance is restricted to these three lines:
> 
> ```
> 		Foo foo = new Foo();
> 		x.add(foo);
> 		foo.bar();
> ```
> 
> But in the code I don't really do anything with that `Foo` instance other than `println` it (i.e., `.toString()`) later on. So I guess the instance's lifetime extends all the way up to the last `println` in `main`.

Sorry for the confusion. By the data-flow I meant the *static* data-flow that Boomerang returns for your query (ForwardBoomerangResults). Which Statement and Val (=Variables) are in the result set? For instance does `Val` `l1` reach the statement `virtualinvoke l1.<Foo: void bar()>();`? If not, there must be an issue with the scene setup.

You described the instance's lifetime which as you write reaches the end of `main`, as the object instance lives as long as the `LinkedList` lives. The static data-flow should compute this information as well, if the soot scene is correctly set up (for instance the code of `LinkedList.add(..)` and `LinkedList.toString()` must be analyzed as well.

**Comment by copumpkin** [comment]
Thanks! I'm going to give you possibly more information than you asked for because I'm not entirely sure how to interpret it yet:

>  The first thing to check is whether the call graph has been computed correctly. Are the edges to bar and to Foo.<init> present?

I'm using a `new SootCallGraph()` in my transformer, and if I list all edges in it for the sample code above, I get the following:

```
Call Graph Edge: CS: l0.<init>() calls <java.lang.Object: void <init>()>
Call Graph Edge: CS: $stack3.<init>() calls <Foo: void <init>()>
Call Graph Edge: CS: registerNatives() calls <java.lang.Object: void <clinit>()>
Call Graph Edge: CS: l1.bar() calls <Foo: void bar()>
```

And here's the complete dump of the units in my method body, just for reference:

```
    public static java.util.List foos()
    {
        java.util.LinkedList $stack2, l0;
        Foo $stack3, l1;
        nop;
        nop;
        $stack2 = new java.util.LinkedList;
        specialinvoke $stack2.<java.util.LinkedList: void <init>()>();
        l0 = $stack2;
        $stack3 = new Foo;
        specialinvoke $stack3.<Foo: void <init>()>();
        l1 = $stack3;
        interfaceinvoke l0.<java.util.List: boolean add(java.lang.Object)>(l1);
        virtualinvoke l1.<Foo: void bar()>();
        return l0;
    }
```

So if I'm interpreting it correctly, I see two edges in that call graph for the `Foo` calls.

> Sorry for the confusion. By the data-flow I meant the static data-flow that Boomerang returns for your query (ForwardBoomerangResults). Which Statement and Val (=Variables) are in the result set? For instance does Val l1 reach the statement virtualinvoke l1.<Foo: void bar()>();? If not, there must be an issue with the scene setup.

Are you asking for the value returned from `.getInvokedMethodOnInstance`? That's the empty set I described earlier. If I call `.getVisitedMethods` my `ForwardBoomerangResults`, I get the following:

```
[<Test: java.util.List foos()>, <Foo: void bar()>, <Foo: void <init>()>, <java.lang.Object: void <init>()>]
```

So it seems like somehow my forward query visited `Foo.bar()`, but didn't include that in my `.getInvokedMethodOnInstance` results?

**Comment by copumpkin** [comment]
Actually, looking at that call graph more carefully, what jumps out at me is that the `List` methods _aren't_ on it, but the `Foo` methods are. This may just be a coincidence, but are the `List` methods appearing in `.getInvokedMethodOnInstance` precisely because they aren't in my call graph?

**Comment by copumpkin** [comment]
Aha, after further investigation I think you're right. I ran `PackManager.v().getPack("cg").apply()` (which I guess I'd missed somehow!) and now my calls are showing up! Thanks for the help!

**Comment by johspaeth** [comment]
Ah great seeing the issue solved.

Despite the fact, it would be good to extend `.getInvokedMethodOnInstance` to also track callsites where edges in the call graph are missing (as it was in your case when `List` was not on the classpath). I have to take a look at it. Thanks for pointing it out, though.

**Comment by copumpkin** [comment]
@johspaeth I'm actually still a bit confused, sorry! As I said, the `cg` did help, but I'm still unclear on what's happening and how include/exclude works with it. If for example, I follow some of your examples and do something like

```
Options.v.set_include(List("java.lang.*", "java.util.*")
```

The behavior I described earlier in the thread comes back and my `List` method invocations don't show up in `.getInvokedMethodOnInstance`. On the other hand, if I add my classes to `set_exclude`, then I start getting all the methods invoked on them. How do those include/exclude options affect this output? Is what I'm describing expected? If not, I've actually reduced my harness code quite a bit (though as with most soot things it's still quite noisy) so I can put up an actual repro.
